x-pack/plugins/security_solution/public/resolver/view/panels/panel_content_related_list.tsx

```diff
@@ -0,0 +1,247 @@
+
+  const waitCrumbs = useMemo(() => {
+    return [
+      {
+        text: eventsString,
+        onClick: () => {
+          pushToQueryParams({ crumbId: '', crumbEvent: '' });
+        },
+      },
+    ];
+  }, [pushToQueryParams, eventsString]);
+
+  const relatedEventsToDisplay = useMemo(() => {
+    return relatedEventsForThisProcess?.events || [];
+  }, [relatedEventsForThisProcess?.events]);
+
+  /**
+   * A list entry will be displayed for each of these
+   */
+  const matchingEventEntries: MatchingEventEntry[] = useMemo(() => {
+    const relateds = relatedEventsToDisplay
+      .reduce((a: ResolverEvent[], candidate) => {
+        if (event.primaryEventCategory(candidate) === eventType) {
+          a.push(candidate);
+        }
+        return a;
+      }, [])
+      .map((resolverEvent) => {
+        const eventTime = event.eventTimestamp(resolverEvent);
+        const formattedDate = typeof eventTime === 'undefined' ? '' : formatDate(eventTime);
+        const entityId = event.eventId(resolverEvent);
+
+        return {
+          formattedDate,
+          eventCategory: `${eventType}`,
+          eventType: `${event.ecsEventType(resolverEvent)}`,
+          name: event.descriptiveName(resolverEvent),
+          entityId,
+          setQueryParams: () => {
+            pushToQueryParams({ crumbId: entityId, crumbEvent: processEntityId });
+          },
+        };
+      });
+    return relateds;
+  }, [relatedEventsToDisplay, eventType, processEntityId, pushToQueryParams]);
+
+  const crumbs = useMemo(() => {
+    return [
+      {
+        text: eventsString,
+        onClick: () => {
+          pushToQueryParams({ crumbId: '', crumbEvent: '' });
+        },
+      },
+      {
+        text: processName,
+        onClick: () => {
+          pushToQueryParams({ crumbId: processEntityId, crumbEvent: '' });
+        },
+      },
+      {
+        text: (
+          <>
+            <FormattedMessage
+              id="xpack.securitySolution.enpoint.resolver.panel.relatedEventList.numberOfEvents"
+              values={{ totalCount }}
+              defaultMessage="{totalCount} Events"
+            />
+          </>
+        ),
+        onClick: () => {
+          pushToQueryParams({ crumbId: processEntityId, crumbEvent: 'all' });
+        },
+      },
+      {
+        text: (
+          <>
+            <FormattedMessage
+              id="xpack.securitySolution.enpoint.resolver.panel.relatedEventList.countByCategory"
+              values={{ count: matchingEventEntries.length, category: eventType }}
+              defaultMessage="{count} {category}"
+            />
+          </>
+        ),
+        onClick: () => {},
+      },
+    ];
+  }, [
+    eventType,
+    eventsString,
+    matchingEventEntries.length,
+    processEntityId,
+    processName,
+    pushToQueryParams,
+    totalCount,
+  ]);
+
+  /**
+   * Wait here until the effect resolves...
+   */
+  if (!relatedsReady) {

```

Comment:
Note for future refactoring work: Let's try to move this conditional up a level.